gadgets/trace_open/test/trace_open_test.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019-2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	gadgetTesting "github.com/inspektor-gadget/inspektor-gadget/gadgets/testing"
+	igtesting "github.com/inspektor-gadget/inspektor-gadget/pkg/testing"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/containers"
+	igrunner "github.com/inspektor-gadget/inspektor-gadget/pkg/testing/ig"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/match"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+type traceOpenEvent struct {
+	eventtypes.CommonData
+
+	MountNsID uint64 `json:"mountnsid"`
+	Pid       uint32 `json:"pid"`
+	Uid       uint32 `json:"uid"`
+	Gid       uint32 `json:"gid"`
+	Comm      string `json:"comm"`
+	Ret       int    `json:"ret"`
+	Flags     int    `json:"flags"`
+	Mode      int    `json:"mode"`
+	FName     string `json:"fname"`
+}
+
+func TestTraceOpen(t *testing.T) {
+	gadgetTesting.RequireEnvironmentVariables(t)
+
+	cn := "test-trace-open"
+	containerFactory, err := containers.NewContainerFactory("docker")
+	require.NoError(t, err, "new container factory")
+
+	traceOpenCmd := igrunner.New(
+		"trace_open",
+		igrunner.WithPathFromEnvVar(),
+		igrunner.WithFlags("--runtimes=docker"),
+		igrunner.WithStartAndStop(),
+		igrunner.WithValidateOutput(
+			func(t *testing.T, output string) {
+				expectedEntry := &traceOpenEvent{
+					CommonData: eventtypes.CommonData{
+						Runtime: eventtypes.BasicRuntimeMetadata{
+							RuntimeName:   eventtypes.String2RuntimeName("docker"),
+							ContainerName: cn,
+						},
+					},
+					Comm:  "cat",
+					Ret:   3,
+					FName: "/dev/null",
+					Uid:   1000,
+					Gid:   1111,
+					Flags: 0,
+					Mode:  0,
+				}
+
+				normalize := func(e *traceOpenEvent) {
+					e.MountNsID = 0
+					e.Pid = 0
+
+					e.Runtime.ContainerID = ""
+					e.Runtime.ContainerImageName = ""
+					e.Runtime.ContainerImageDigest = ""
+				}
+
+				match.ExpectEntriesToMatch(t, output, normalize, expectedEntry)
+			}),
+	)
+
+	testSteps := []igtesting.TestStep{
+		traceOpenCmd,
+		containerFactory.NewContainer(cn, "setuidgid 1000:1111 cat /dev/null"),
```

Comment:
Remember to ensure this command is executed multiple times with the `while` I suggested [here](https://github.com/inspektor-gadget/inspektor-gadget/pull/2607#discussion_r1556404269). Otherwise, you risk missing the event because ig is pulling the image.